File: src/_opensomeip/exceptions.cpp
#include <pybind11/pybind11.h>
#include <stdexcept>

namespace py = pybind11;

static py::exception<std::runtime_error> someip_error;
static py::exception<std::runtime_error> transport_error;
static py::exception<std::runtime_error> connection_error;
static py::exception<std::runtime_error> connection_lost_error;
static py::exception<std::runtime_error> serialization_error;
static py::exception<std::runtime_error> timeout_error;
static py::exception<std::runtime_error> sd_error;
static py::exception<std::runtime_error> rpc_error;
static py::exception<std::runtime_error> e2e_error;
static py::exception<std::runtime_error> config_error;

void init_exceptions(py::module_& m) {
    someip_error = py::register_exception<std::runtime_error>(m, "SomeIpError");
    transport_error = py::register_exception<std::runtime_error>(
        m, "TransportError", someip_error);
    connection_error = py::register_exception<std::runtime_error>(
        m, "ConnectionError", transport_error);
    connection_lost_error = py::register_exception<std::runtime_error>(
        m, "ConnectionLostError", transport_error);
    serialization_error = py::register_exception<std::runtime_error>(
        m, "SerializationError", someip_error);
    timeout_error = py::register_exception<std::runtime_error>(
        m, "TimeoutError", someip_error);
    sd_error = py::register_exception<std::runtime_error>(
        m, "ServiceDiscoveryError", someip_error);
    rpc_error = py::register_exception<std::runtime_error>(
        m, "RpcError", someip_error);
    e2e_error = py::register_exception<std::runtime_error>(
        m, "E2EError", someip_error);
    config_error = py::register_exception<std::runtime_error>(
        m, "ConfigurationError", someip_error);
}

File: src/opensomeip/exceptions.py
"""Exception hierarchy for opensomeip.

All exceptions inherit from :class:`SomeIpError`. This module is pure Python
so it can be imported for type-checking even without the compiled extension.
"""

from __future__ import annotations


class SomeIpError(Exception):
    """Base exception for all opensomeip errors."""


class TransportError(SomeIpError):
    """Error related to SOME/IP transport operations."""


class ConnectionError(TransportError):
    """Failed to establish a transport connection."""


class ConnectionLostError(TransportError):
    """An established transport connection was lost."""


class SerializationError(SomeIpError):
    """Error during SOME/IP payload serialization or deserialization."""


class TimeoutError(SomeIpError):
    """A SOME/IP operation timed out."""


class ServiceDiscoveryError(SomeIpError):
    """Error related to SOME/IP Service Discovery."""


class RpcError(SomeIpError):
    """Error during a SOME/IP RPC call."""


class E2EError(SomeIpError):
    """Error related to E2E (End-to-End) protection."""


class ConfigurationError(SomeIpError):
    """Invalid configuration provided to an opensomeip component."""
